fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/codec/LuceneOptimizedFieldInfosFormat.java

```diff
@@ -20,36 +20,277 @@
+        if (rawBytes == null) {
+            throw new FileNotFoundException("Could not find field info");
+        }
+        return LuceneFieldInfosProto.FieldInfos.newBuilder().mergeFrom(rawBytes);
     }
 
     @Override
     public void write(final Directory directory, final SegmentInfo segmentInfo, final String segmentSuffix, final FieldInfos infos, final IOContext context) throws IOException {
-        fieldInfosFormat.write(new LuceneOptimizedWrappedDirectory(directory, infos), segmentInfo, segmentSuffix, infos, context);
+        // Bitset to track what fields in the global info we are using
+        // we still save this even if not reusing global, in case we later reuse field infos other than the global one
+        final BitSet bitSet = new BitSet();
+        final LuceneFieldInfosProto.FieldInfos.Builder protobuf = luceneToProto(infos, bitSet);
+        @SuppressWarnings("PMD.CloseResource") // we don't need to close this because it is just extracting from the directory
+        final FDBDirectory fdbDirectory = getFdbDirectory(directory);
+        // It would probably be valuable to cache the parsed global field infos proto rather than re-parsing for every
+        // segment
+        final byte[] rawGlobalBytes = fdbDirectory.readGlobalFieldInfos();
+        boolean globalNeedsUpdating = false;
+        boolean canReuseGlobal = true;
+        final LuceneFieldInfosProto.FieldInfos.Builder globalFieldInfos;
+        if (rawGlobalBytes == null) {
+            globalNeedsUpdating = true;
+            globalFieldInfos = protobuf;
+        } else {
+            globalFieldInfos = parseFieldInfo(rawGlobalBytes);
+            final Map<Integer, LuceneFieldInfosProto.FieldInfo> globalFieldInfo = globalFieldInfos.getFieldInfoList().stream()
+                    .collect(Collectors.toMap(LuceneFieldInfosProto.FieldInfo::getNumber, Function.identity()));
+            for (final LuceneFieldInfosProto.FieldInfo fieldInfo : protobuf.getFieldInfoList()) {
+                final LuceneFieldInfosProto.FieldInfo globalVersion = globalFieldInfo.get(fieldInfo.getNumber());
+                if (globalVersion == null) {
+                    // new field ... we can add it
+                    globalFieldInfos.addFieldInfo(fieldInfo);
+                    globalNeedsUpdating = true;
+                } else {
+                    if (!globalVersion.equals(fieldInfo)) {
+                        canReuseGlobal = false;
+                        break;
+                    } // The field is already in the global, and we can continue to reuse
+                }
+            }
+        }
+        final long id;
+        if (!canReuseGlobal) {
+            id = fdbDirectory.writeFieldInfo(protobuf.build().toByteArray());
+        } else {
+            id = FDBDirectory.GLOBAL_FIELD_INFOS_ID;
+        }
+        if (globalNeedsUpdating) {
```

Comment:
It is possible, by the algorithm above, to have both `canReuseGlobal` and `globalNeedsUpdating` as `TRUE`, in which case we will write both the custom version and the revised global versions to the DB. Is that intended?